ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/ComparisonInterpreter.java

```diff
@@ -0,0 +1,222 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.interpreter;
+
+import io.confluent.ksql.execution.expression.tree.ComparisonExpression;
+import io.confluent.ksql.execution.interpreter.CastInterpreter.ConversionType;
+import io.confluent.ksql.execution.interpreter.CastInterpreter.NumberConversions;
+import io.confluent.ksql.execution.interpreter.terms.BasicTerms.BooleanTerm;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.CompareToTerm;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.ComparisonCheckFunction;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.ComparisonFunction;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.ComparisonNullCheckFunction;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.EqualsCheckFunction;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.EqualsFunction;
+import io.confluent.ksql.execution.interpreter.terms.ComparisonTerm.EqualsTerm;
+import io.confluent.ksql.execution.interpreter.terms.Term;
+import io.confluent.ksql.schema.ksql.SqlTimestamps;
+import io.confluent.ksql.schema.ksql.types.SqlBaseType;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.KsqlException;
+import java.math.BigDecimal;
+import java.sql.Timestamp;
+import java.util.Optional;
+
+/**
+ * Contains lots of the logic for doing comparisons in the interpreter.
+ */
+public final class ComparisonInterpreter {
+
+  private ComparisonInterpreter() { }
+
+  /**
+   * Does a comparison between the two terms
+   * @param type The type of the comparison
+   * @param left Left term
+   * @param right Right term
+   * @return The term representing the result of the comparison
+   */
+  public static BooleanTerm doComparison(
+      final ComparisonExpression.Type type,
+      final Term left,
+      final Term right
+  ) {
+    final ComparisonNullCheckFunction nullCheckFunction = getNullCheckFunction(type);
+
+    final Optional<ComparisonFunction> compareTo = doCompareTo(left, right);
+    if (compareTo.isPresent()) {
+      return new CompareToTerm(left, right, nullCheckFunction, compareTo.get(),
+          getComparisonCheckFunction(type, left, right));
+    }
+
+    final Optional<EqualsFunction> equals = doEquals(left, right);
+    if (equals.isPresent()) {
+      return new EqualsTerm(left, right, nullCheckFunction, equals.get(),
+          getEqualsCheckFunction(type, left, right));
+    }
+
+    throw new KsqlException("Unsupported comparison between " + left.getSqlType() + " and "
+        + right.getSqlType());
+  }
+
+  private static BigDecimal toDecimal(final Object object, final SqlType from,
```

Comment:
can we leverage the cast functionality here (and below)?

Reply:
One utilizes implicit conversion and the other explicit where you must provide precision and scale.  The implicit methods are used in the java code gen version and I wanted to make sure I had parity with it.  The precision and scale are chosen automatically in such a way that they fit the number, unlike the cast version.

Reply:
ack, good to know